### PR TITLE
Use signature based format detection

### LINKED ISSUE #2365: Loading savegame don't work with address sanitizers and xml format
### Environment
* **FreeOrion Version**: 3406e4e
* **Operating System**: Gentoo Linux AMD64
* **Graphic card used**: NVIDIA GeForce GTX 750 Ti
* **Fetched as** 
  * Compiled from source

### Description
When I trying to use address sanitizer I get error when trying to load game. Instead of throwing exception sanitizer aborts the server.
I suppose it better to use same way already used for network messages to check if file starts from '<?xml'.

```
$ ASAN_OPTIONS=detect_leaks=0 ./bin/freeorion
==19415==WARNING: AddressSanitizer failed to allocate 0x6576206c6d783f3d bytes
==19415==AddressSanitizer's allocator is terminating the process instead of returning 0
==19415==If you don't like this behavior set allocator_may_return_null=1
==19415==AddressSanitizer CHECK failed: /tmp/portage/sys-devel/gcc-7.3.0-r3/work/gcc-7.3.0/libsanitizer/sanitizer_common/sanitizer_allocator.cc:218 "((0)) != (0)" (0x0, 0x0)
    #0 0x7f2dad87add2  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0xe9dd2)
    #1 0x7f2dad8996a5 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0x1086a5)
    #2 0x7f2dad880532  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0xef532)
    #3 0x7f2dad7b9ce6  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0x28ce6)
    #4 0x7f2dad8715fe in operator new(unsigned long) (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0xe05fe)
    #5 0x7f2da773ef69 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_mutate(unsigned long, unsigned long, char const*, unsigned long) (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libstdc++.so.6+0x152f69)
    #6 0x7f2da773f96f in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_replace_aux(unsigned long, unsigned long, unsigned long, char) (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libstdc++.so.6+0x15396f)
    #7 0x7f2da804d51f in boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (/usr/lib64/libboost_serialization.so.1.65.0+0x1a51f)
    #8 0x7f2da804d71a in boost::archive::basic_binary_iarchive<boost::archive::binary_iarchive>::init() (/usr/lib64/libboost_serialization.so.1.65.0+0x1a71a)
    #9 0x200000904bf0 in boost::archive::binary_iarchive_impl<boost::archive::binary_iarchive, char, std::char_traits<char> >::binary_iarchive_impl(std::istream&, unsigned int) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x200000904bf0)
    #10 0x2000008cdae5 in boost::archive::binary_iarchive::binary_iarchive(std::istream&, unsigned int) /usr/include/boost/archive/binary_iarchive.hpp:44
    #11 0x2000008cdae5 in LoadPlayerSaveHeaderData(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<PlayerSaveHeaderData, std::allocator<PlayerSaveHeaderData> >&) /mnt/another/srcs/GIT/freeorion/server/SaveLoad.cpp:506
    #12 0x200000ad471b in GetHostNameFromSinglePlayerSetupData /mnt/another/srcs/GIT/freeorion/server/ServerFSM.cpp:94
    #13 0x200000b293e7 in Idle::react(HostSPGame const&) /mnt/another/srcs/GIT/freeorion/server/ServerFSM.cpp:507
    #14 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::custom_reaction<HostSPGame>::react<Idle, boost::statechart::event_base, void const*>(Idle&, boost::statechart::event_base const&, void const* const&) /usr/include/boost/statechart/custom_reaction.hpp:41
    #15 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react_impl_non_empty::local_react_impl<boost::mpl::list4<boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless> >, boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0> >(boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>&, boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:813
    #16 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react<boost::mpl::list4<boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless> > >(boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:851
    #17 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react_impl_non_empty::local_react_impl<boost::mpl::list<boost::statechart::custom_reaction<HostMPGame>, boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0> >(boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>&, boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:820
    #18 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react<boost::mpl::list<boost::statechart::custom_reaction<HostMPGame>, boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na> >(boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:851
    #19 0x200000bb7e6a in boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::react_impl(boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:488
    #20 0x200000a980fe in boost::statechart::state_machine<ServerFSM, Idle, std::allocator<void>, boost::statechart::null_exception_translator>::send_event(boost::statechart::event_base const&) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x200000a980fe)
    #21 0x200000a99a2e in boost::statechart::state_machine<ServerFSM, Idle, std::allocator<void>, boost::statechart::null_exception_translator>::process_event(boost::statechart::event_base const&) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x200000a99a2e)
    #22 0x2000009f2aaf in ServerApp::HandleNonPlayerMessage(Message const&, std::shared_ptr<PlayerConnection>) /mnt/another/srcs/GIT/freeorion/server/ServerApp.cpp:448
    #23 0x200000a5353b in boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >::operator()(ServerApp*, Message const&, std::shared_ptr<PlayerConnection>) const /usr/include/boost/bind/mem_fn_template.hpp:280
    #24 0x200000a5353b in void boost::_bi::list3<boost::_bi::value<ServerApp*>, boost::arg<1>, boost::arg<2> >::operator()<boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >, boost::_bi::rrlist2<Message, std::shared_ptr<PlayerConnection> > >(boost::_bi::type<void>, boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >&, boost::_bi::rrlist2<Message, std::shared_ptr<PlayerConnection> >&, int) /usr/include/boost/bind/bind.hpp:398
    #25 0x200000a5353b in void boost::_bi::bind_t<void, boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >, boost::_bi::list3<boost::_bi::value<ServerApp*>, boost::arg<1>, boost::arg<2> > >::operator()<Message, std::shared_ptr<PlayerConnection> >(Message&&, std::shared_ptr<PlayerConnection>&&) /usr/include/boost/bind/bind.hpp:1318
    #26 0x200000a5353b in boost::detail::function::void_function_obj_invoker2<boost::_bi::bind_t<void, boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >, boost::_bi::list3<boost::_bi::value<ServerApp*>, boost::arg<1>, boost::arg<2> > >, void, Message, std::shared_ptr<PlayerConnection> >::invoke(boost::detail::function::function_buffer&, Message, std::shared_ptr<PlayerConnection>) /usr/include/boost/function/function_template.hpp:159
    #27 0x2000002111c4 in boost::detail::function::void_function_obj_invoker0<boost::_bi::bind_t<boost::_bi::unspecified, boost::function<void (Message, std::shared_ptr<PlayerConnection>)>, boost::_bi::list2<boost::_bi::value<Message>, boost::_bi::value<std::shared_ptr<PlayerConnection> > > >, void>::invoke(boost::detail::function::function_buffer&) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x2000002111c4)
    #28 0x2000001bbd7a in boost::function0<void>::operator()() const /usr/include/boost/function/function_template.hpp:759
    #29 0x2000001bbd7a in ServerNetworking::HandleNextEvent() /mnt/another/srcs/GIT/freeorion/network/ServerNetworking.cpp:612
    #30 0x200000975329 in ServerApp::Run() /mnt/another/srcs/GIT/freeorion/server/ServerApp.cpp:302
    #31 0x200000976b2d in ServerApp::operator()() /mnt/another/srcs/GIT/freeorion/server/ServerApp.cpp:129
    #32 0x20000089dac7 in main /mnt/another/srcs/GIT/freeorion/server/dmain.cpp:89
    #33 0x7f2da6a85ae6 in __libc_start_main (/lib64/libc.so.6+0x21ae6)
    #34 0x2000000f13e9 in _start (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x2000000f13e9)

$ ASAN_OPTIONS=detect_leaks=0:allocator_may_return_null=1 ./bin/freeorion
==22898==WARNING: AddressSanitizer failed to allocate 0x6576206c6d783f3d bytes
=================================================================
==22898==AddressSanitizer CHECK failed: /tmp/portage/sys-devel/gcc-7.3.0-r3/work/gcc-7.3.0/libsanitizer/asan/asan_descriptions.cc:80 "((0 && "Address is not in memory and not in shadow?")) != (0)" (0x0, 0x0)
    #0 0x7f2a1407add2  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0xe9dd2)
    #1 0x7f2a140996a5 in __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0x1086a5)
    #2 0x7f2a13fbe77c  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0x2d77c)
    #3 0x7f2a13fbf842  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0x2e842)
    #4 0x7f2a13fc1abf  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0x30abf)
    #5 0x7f2a1407a504  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0xe9504)
    #6 0x7f2a13fefdbb  (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libasan.so.4+0x5edbb)
    #7 0x7f2a0df3f939 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_replace_aux(unsigned long, unsigned long, unsigned long, char) (/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/libstdc++.so.6+0x153939)
    #8 0x7f2a0e84d51f in boost::archive::basic_binary_iprimitive<boost::archive::binary_iarchive, char, std::char_traits<char> >::load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (/usr/lib64/libboost_serialization.so.1.65.0+0x1a51f)
    #9 0x7f2a0e84d71a in boost::archive::basic_binary_iarchive<boost::archive::binary_iarchive>::init() (/usr/lib64/libboost_serialization.so.1.65.0+0x1a71a)
    #10 0x200000904bf0 in boost::archive::binary_iarchive_impl<boost::archive::binary_iarchive, char, std::char_traits<char> >::binary_iarchive_impl(std::istream&, unsigned int) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x200000904bf0)
    #11 0x2000008cdae5 in boost::archive::binary_iarchive::binary_iarchive(std::istream&, unsigned int) /usr/include/boost/archive/binary_iarchive.hpp:44
    #12 0x2000008cdae5 in LoadPlayerSaveHeaderData(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<PlayerSaveHeaderData, std::allocator<PlayerSaveHeaderData> >&) /mnt/another/srcs/GIT/freeorion/server/SaveLoad.cpp:506
    #13 0x200000ad471b in GetHostNameFromSinglePlayerSetupData /mnt/another/srcs/GIT/freeorion/server/ServerFSM.cpp:94
    #14 0x200000b293e7 in Idle::react(HostSPGame const&) /mnt/another/srcs/GIT/freeorion/server/ServerFSM.cpp:507
    #15 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::custom_reaction<HostSPGame>::react<Idle, boost::statechart::event_base, void const*>(Idle&, boost::statechart::event_base const&, void const* const&) /usr/include/boost/statechart/custom_reaction.hpp:41
    #16 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react_impl_non_empty::local_react_impl<boost::mpl::list4<boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless> >, boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0> >(boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>&, boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:813
    #17 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react<boost::mpl::list4<boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless> > >(boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:851
    #18 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react_impl_non_empty::local_react_impl<boost::mpl::list<boost::statechart::custom_reaction<HostMPGame>, boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0> >(boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>&, boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:820
    #19 0x200000bb7e6a in boost::statechart::detail::reaction_result boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::local_react<boost::mpl::list<boost::statechart::custom_reaction<HostMPGame>, boost::statechart::custom_reaction<HostSPGame>, boost::statechart::custom_reaction<ShutdownServer>, boost::statechart::custom_reaction<Error>, boost::statechart::custom_reaction<Hostless>, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na> >(boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:851
    #20 0x200000bb7e6a in boost::statechart::simple_state<Idle, ServerFSM, boost::mpl::list<mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>, (boost::statechart::history_mode)0>::react_impl(boost::statechart::event_base const&, void const*) /usr/include/boost/statechart/simple_state.hpp:488
    #21 0x200000a980fe in boost::statechart::state_machine<ServerFSM, Idle, std::allocator<void>, boost::statechart::null_exception_translator>::send_event(boost::statechart::event_base const&) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x200000a980fe)
    #22 0x200000a99a2e in boost::statechart::state_machine<ServerFSM, Idle, std::allocator<void>, boost::statechart::null_exception_translator>::process_event(boost::statechart::event_base const&) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x200000a99a2e)
    #23 0x2000009f2aaf in ServerApp::HandleNonPlayerMessage(Message const&, std::shared_ptr<PlayerConnection>) /mnt/another/srcs/GIT/freeorion/server/ServerApp.cpp:448
    #24 0x200000a5353b in boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >::operator()(ServerApp*, Message const&, std::shared_ptr<PlayerConnection>) const /usr/include/boost/bind/mem_fn_template.hpp:280
    #25 0x200000a5353b in void boost::_bi::list3<boost::_bi::value<ServerApp*>, boost::arg<1>, boost::arg<2> >::operator()<boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >, boost::_bi::rrlist2<Message, std::shared_ptr<PlayerConnection> > >(boost::_bi::type<void>, boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >&, boost::_bi::rrlist2<Message, std::shared_ptr<PlayerConnection> >&, int) /usr/include/boost/bind/bind.hpp:398
    #26 0x200000a5353b in void boost::_bi::bind_t<void, boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >, boost::_bi::list3<boost::_bi::value<ServerApp*>, boost::arg<1>, boost::arg<2> > >::operator()<Message, std::shared_ptr<PlayerConnection> >(Message&&, std::shared_ptr<PlayerConnection>&&) /usr/include/boost/bind/bind.hpp:1318
    #27 0x200000a5353b in boost::detail::function::void_function_obj_invoker2<boost::_bi::bind_t<void, boost::_mfi::mf2<void, ServerApp, Message const&, std::shared_ptr<PlayerConnection> >, boost::_bi::list3<boost::_bi::value<ServerApp*>, boost::arg<1>, boost::arg<2> > >, void, Message, std::shared_ptr<PlayerConnection> >::invoke(boost::detail::function::function_buffer&, Message, std::shared_ptr<PlayerConnection>) /usr/include/boost/function/function_template.hpp:159
    #28 0x2000002111c4 in boost::detail::function::void_function_obj_invoker0<boost::_bi::bind_t<boost::_bi::unspecified, boost::function<void (Message, std::shared_ptr<PlayerConnection>)>, boost::_bi::list2<boost::_bi::value<Message>, boost::_bi::value<std::shared_ptr<PlayerConnection> > > >, void>::invoke(boost::detail::function::function_buffer&) (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x2000002111c4)
    #29 0x2000001bbd7a in boost::function0<void>::operator()() const /usr/include/boost/function/function_template.hpp:759
    #30 0x2000001bbd7a in ServerNetworking::HandleNextEvent() /mnt/another/srcs/GIT/freeorion/network/ServerNetworking.cpp:612
    #31 0x200000975329 in ServerApp::Run() /mnt/another/srcs/GIT/freeorion/server/ServerApp.cpp:302
    #32 0x200000976b2d in ServerApp::operator()() /mnt/another/srcs/GIT/freeorion/server/ServerApp.cpp:129
    #33 0x20000089dac7 in main /mnt/another/srcs/GIT/freeorion/server/dmain.cpp:89
    #34 0x7f2a0d285ae6 in __libc_start_main (/lib64/libc.so.6+0x21ae6)
    #35 0x2000000f13e9 in _start (/mnt/new/o01eg/games/freeorion-debug/bin/freeoriond+0x2000000f13e9)
```

### Expected Result
Server successfully loads game.

### Steps to reproduce
* Build freeorion with address sanitizer
* Try to load savegame in xml or xml-zlib format.

### PR DESCRIPTION
**Pull request by o01eg**
Fixes behaviour with strict allocation rules enforced by sanitizers.

Fixes #2365

### CODE REVIEW

File: server/SaveLoad.cpp [hunk 1]
```diff
@@ -333,8 +333,12 @@ void LoadGame(const std::string& filename, ServerSaveGameData& server_save_game_
...
 
-        try {
-            // first attempt binary deserialziation
+        std::string signature(5, '\0');
+        if (!ifs.read(&signature[0], 5))
+            throw std::runtime_error(UNABLE_TO_OPEN_FILE);
+        boost::iostreams::seek(ifs, 0, std::ios_base::beg);
+
+        if (strncmp(signature.c_str(), "<?xml", 5)) {
             freeorion_bin_iarchive ia(ifs);
```
Comment (geoffthemedio) [inline server/SaveLoad.cpp:343]: would be clearer / helpful to have a comment here like
`// XML file format signature not found; try as binary`

File: server/SaveLoad.cpp [hunk 2]
```diff
@@ -498,29 +497,29 @@ void LoadPlayerSaveHeaderData(const std::string& filename, std::vector<PlayerSav
...
             ia >> BOOST_SERIALIZATION_NVP(player_save_header_data);
-
-        } catch (...) {
-            // if binary deserialization failed, try more-portable XML deserialization
-            DebugLogger() << "Trying again with XML deserialization...";
-
-            // reset to start of stream (attempted binary serialization will have consumed some input...)
-            boost::iostreams::seek(ifs, 0, std::ios_base::beg);
+        } else {
+            DebugLogger() << "Attempting XML deserialization...";
```
Comment (geoffthemedio) [inline server/SaveLoad.cpp:520]: why a log message here but not other similar places?

**Comment by geoffthemedio** [comment]
otherwise looks fine

File: server/SaveLoad.cpp (hunk 1 above)
Comment (o01eg) [reply server/SaveLoad.cpp:343]: Added to all places.

File: server/SaveLoad.cpp (hunk 2 above)
Comment (o01eg) [reply server/SaveLoad.cpp:520]: Made log messages symmetrical.

**Comment by o01eg** [comment]
Rebased. Fixed.

**Comment by subpop** [comment]
Sure does! I'll try to backport that patch.

**Pull request merged**